btx/processing/rawimagetimetool.py

```diff
@@ -0,0 +1,575 @@
+        self.write_file(self.edges_fit, fname, outdir)
+        ## Fitted delays
+        fname = f'DelaysFit_{run}.out'
+        self.write_file(self.delays_fit, fname, outdir)
+
+        if figs:
+            self.plot_calib(delays, edges, self._model)
+            self.plot_hist(self.edges_fit)
+
+    def format_run(self) -> str:
+        """! Format the run(s) number for output file names. The format is
+        r0000 for a single run and r0000-r0000 for a range of runs.
+        """
+        run = self.ds.ds_string.split(':')[1].split('=')[1]
+        if '-' in run:
+            run = run.split('-')
+            start = run[0]
+            end = run[1]
+
+            run = f'r{int(start):04}-{int(end):04}'
+        else:
+            run = f'r{int(run):04}'
+        return run
+
+    def process_run(self, calib=True) -> (np.array, np.array, np.array):
+        """! Perform edge detection for all images in a run.
+
+        @param calib (bool) Whether or not this is a calibration run.
+
+        @return delays (np.array) Array of delays used.
+        @return edges (np.array) Detected edge position.
+        @return ampls (np.array) Convolution amplitudes for the detected edges.
+        @return stamps (np.array) Unique event identifier stamps. NOT returned in calibration.
+        """
+        ## @todo Split edge detection into separate function from loop for ease
+        # of use with jitter correction once model is known.
+
+        # Create kernel for convolution and edge detection.
+        kernel = np.zeros([300])
+        kernel[:150] = 1
+
+        delays = []
+        edges = []
+        ampls = [] # Can be used for filtering good edges from bad
+        fwhms = [] # Can be used for filtering good edges from bad
+
+        if not calib:
+            stamps = []
+
+        stage_code = self.ttstage_code(self.hutch)
+
+        # Loop through run events
+        for idx, evt in enumerate(self.ds.events()):
+            delays.append(self.ds.env().epicsStore().value(stage_code))
+            try:
+                # Retrieve time tool camera image for the event
+                img = self.det.image(evt=evt)
+                edge, ampl, fwhm = self.detect_edge(img, kernel)
+
+                edges.append(edge)
+                ampls.append(ampl)
+                fwhms.append(fwhm)
+
+                if not calib:
+                    evtid = evt.get(psana.EventId)
+                    evtfid = evtid.fiducials()
+                    evttime = evtid.time()
+                    stamp = f'{evttime[0]}-{evttime[1]}-{evtfid}'
+                    stamps.append(stamp)
+
+            except Exception as e:
+                # BAD - Do specific exception handling
+                # If error occurs while getting the image remove the last entry
+                # in the delays list
+                # Errors occur because there are some missing camera images,
+                # but the stage position still registers.
```

Comment:
I agree. Depending on what the answer is to the `delays` question just above, we might want to append to `delays` inside the `try:` block for example?